_posts/2021-03-31-relaxing-requirements-of-moved-from-objects.md

```diff
@@ -139,24 +152,29 @@ for (std::string line; std::getline(std::cin, line);) {
+For the call to `std::getline()` to be valid with a moved-from string, `std::string()` must _guarantee_ that `std::erase()` is valid on a moved-from string. <!-- I don't understand the assertion.  How is erase involved? --> Changing the requirements of a moved-from object does not change the guarantees of the standard components.
 
 ## Impact on the Standard
 
  All components which are _Movable_ in the Standard Library currently satisfy the proposed requirements as stated by both options below. **Both options are non-breaking changes and relax the requirements.** With either option, it may be possible to adopt these options retroactively as part of addressing a defect since neither option is a breaking change.
 
+<!-- Seems likely that the assertion that this is a non-breaking change will be argued with. You simply can't ever loosen or tighten concept requirements without breaking somebody.  If someone wrote an algorithm that depends on the current stronger requirements, that algorithm is now broken with respect to its documentation, and a use of that algorithm with some type that has been documented to satisfy those requirements may break when the implementation of that type changes.  Maybe you want to say “non-code-breaking changes?” -->
+
 Concern has been raised that changing the documentation for requirements, especially the named requirements and concepts, would break existing code documentation that referenced the standard. However, taking a strict view of this would mean that the standard documentation could not be changed. For example, one of the libraries I work on has a `task<>` template which is documented as being ["Similar to `std::function` except it is not copyable and supports move-only and mutable callable targets..."](https://stlab.cc/libraries/concurrency/task/task/). Of note, this goes on to specify, "`stlab::task` satisfies the requirements of [MoveConstructible](https://en.cppreference.com/w/cpp/named_req/MoveConstructible) and [MoveAssignable](https://en.cppreference.com/w/cpp/named_req/MoveAssignable)." Weakening the requirements would mean that statement is still true.
 
 However, the concern over changing the documentation is one that should be considered in light of weakening the requirements retroactively as a defect. As that would mean even citing a specific version of the standard would break. 
 
+<!-- I don't really understand the last sentence, and my sense is that politically, it's better not to try to pull off this line of argumentation, and just make a best effort at assessing the risk.  I think there are some requirements that we absolutely wouldn't consider weakening because the risk is too great.  In fact I am starting to think the better line here is to argue that Geoffrey misinterpreted the standard wording and the non-normative note is wrong, so this doesn't represent an actual weakening of requirements… IIUC  -->
```

Comment:
That would be an interesting argument but many, including Howard Hinnant, agree the intent of the current standard is that a moved-from object must satisfy all the requirements of a non-moved-from object in whatever context it is used. Basically, we'd be arguing Howard didn't intend what he wrote.

Reply:
Whatever his intent, I am pretty sure I’m responsible for the original language, and Howard’s intent conflicts with the intent of that.

Reply:
To elaborate, the fact that it went into a non-normative note is an indication that it was merely meant to clarify the intent of something that was already, strictly-speaking, specified by the standard.  But the  original wording was not intended to say that, and it seems unreasonable to draw the conclusion that that it was, especially in light of the fact that the meaning Howard intended to “clarify” into existence is *unworkable*.  The proper cure is to remove the incorrect non-normative wording, at least as a start.